generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -164,6 +166,157 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
         depth, fuzzer)));
   }
 
+  /**
+   * Function to generate an opaque zero or one by bitwise shifting left or right by an
+   * arbitrary number of bits. Generates an opaque value if the type of zero to be generated is an
+   * integer and the shading language in use supports bitwise operations.
+   *
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from bitwise shifting.
+   */
+  private Optional<Expr> opaqueZeroOrOneBitwiseShift(BasicType type, boolean constContext,
+                                                     final int depth, Fuzzer fuzzer,
+                                                     boolean isZero) {
+    if (!BasicType.allIntegerTypes().contains(type)) {
+      return Optional.empty();
+    }
+    if (!shadingLanguageVersion.supportedBitwiseOperations()) {
+      return Optional.empty();
+    }
+    final BinOp operator = generator.nextBoolean() ? BinOp.SHL : BinOp.SHR;
+    // We need to make sure the LHS and RHS of the shift expression are based on the same type
```

Comment:
The GLSL ES 3.20 spec says:

"One operand can be signed while the other is unsigned. In all cases, the resulting type will be the same type as the left operand."

So I suggest that you randomly choose whether the RHS is signed or unsigned.

Reply:
Also, the spec says:

"If the first operand is a vector, the second operand
must be a scalar or a vector with the same size as the first operand, and the result is computed
component-wise."

So I suggest that you randomly choose whether the second argument should be scalar or vector, such that if it is chosen to be vector it has the same number of elements as the LHS (so that if the LHS is already a scalar, the RHS is a scalar either way).

Reply:
Now that we're clamping, changing the type in the clamp starts to cause issues with applying other identity functions to the resulting expressions (specifically, ternary type mismatches).